nntrainer/layers/mol_attention_layer.cpp

```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
   u_base.add_i(-0.5);
   Tensor u_neg = u_base;
 
-  Tensor integral_left, integral_right;
-  sigmoid.run_fn(u_pos.subtract(state).divide(beta.add(1e-8f)), integral_left);
-  sigmoid.run_fn(u_neg.subtract(state).divide(beta.add(1e-8f)), integral_right);
-  Tensor integral = integral_left.subtract(integral_right);
+  Tensor beta_eps = beta.add(1e-8f);
+
+  Tensor u_pos_m = u_pos.subtract(m);
+  u_pos_m.divide(beta_eps, u_pos_div);
+  sigmoid.run_fn(u_pos_div, prob_left);
 
-  Tensor integral_scaled = integral.multiply(alpha);
-  Tensor scores = integral_scaled.sum(3);
-  scores.reshape(TensorDim({scores.batch(), 1, 1, scores.height()}));
+  Tensor u_neg_m = u_neg.subtract(m);
+  u_neg_m.divide(beta_eps, u_neg_div);
+  sigmoid.run_fn(u_neg_div, prob_right);
+
+  prob_left.subtract(prob_right, prob);
+
+  Tensor prob_scaled = prob.multiply(alpha);
+  prob_scaled.sum(3, scores);
 
   scores.dotBatched(value, output);
 }
 
-void MoLAttentionLayer::calcDerivative(RunLayerContext &context) { /** NYI */
+void MoLAttentionLayer::calcDerivativeHelper(RunLayerContext &context,
+                                             Tensor &dstate) {
+  Tensor &query = context.getInput(wt_idx[AttentionParams::query]);
+  Tensor &value = context.getInput(wt_idx[AttentionParams::value]);
+
+  Tensor &derivative = context.getIncomingDerivative(SINGLE_INOUT_IDX);
+
+  Tensor &fc_proj_out = context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &dfc_proj_out =
+    context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &scores = context.getTensor(wt_idx[AttentionParams::scores]);
+  Tensor &prob = context.getTensor(wt_idx[AttentionParams::prob]);
+  Tensor &prob_left = context.getTensor(wt_idx[AttentionParams::prob_left]);
+  Tensor &prob_right = context.getTensor(wt_idx[AttentionParams::prob_right]);
+  Tensor &u_neg_div = context.getTensor(wt_idx[AttentionParams::u_neg_div]);
```

Comment:
how about make this trainable tensor and use TensorGrad?

Reply:
A lot of tensors are locally allocated in calcDerivativeHelper(). I have a todo to optimize for it and will use this suggestion when optimizing.